src/cmd/services/m3coordinator/downsample/downsampler.go

```diff
@@ -94,34 +100,90 @@ func newDownsampler(opts downsamplerOptions) (*downsampler, error) {
 		return nil, err
 	}
 
+	downsampler := &downsampler{
+		opts:                opts.opts,
+		agg:                 opts.agg,
+		metricsAppenderOpts: defaultMetricsAppenderOptions(opts.opts, opts.agg),
+	}
+
+	// No need to retain watch as NamespaceWatcher.Close() will handle closing any watches
+	// generated by creating listeners.
+	_ = downsampler.opts.ClusterNamespacesWatcher.RegisterListener(downsampler)
+
+	return downsampler, nil
+}
+
+func defaultMetricsAppenderOptions(opts DownsamplerOptions, agg agg) metricsAppenderOptions {
 	debugLogging := false
-	logger := opts.opts.InstrumentOptions.Logger()
+	logger := opts.InstrumentOptions.Logger()
 	if logger.Check(zapcore.DebugLevel, "debug") != nil {
 		debugLogging = true
 	}
 
-	metricsAppenderOpts := metricsAppenderOptions{
-		agg:                          opts.agg.aggregator,
-		clientRemote:                 opts.agg.clientRemote,
-		defaultStagedMetadatasProtos: opts.agg.defaultStagedMetadatasProtos,
-		clockOpts:                    opts.agg.clockOpts,
-		tagEncoderPool:               opts.agg.pools.tagEncoderPool,
-		matcher:                      opts.agg.matcher,
-		metricTagsIteratorPool:       opts.agg.pools.metricTagsIteratorPool,
-		debugLogging:                 debugLogging,
-		logger:                       logger,
-		augmentM3Tags:                opts.agg.m3PrefixFilter,
+	return metricsAppenderOptions{
+		agg:                    agg.aggregator,
+		clientRemote:           agg.clientRemote,
+		clockOpts:              agg.clockOpts,
+		tagEncoderPool:         agg.pools.tagEncoderPool,
+		matcher:                agg.matcher,
+		metricTagsIteratorPool: agg.pools.metricTagsIteratorPool,
+		debugLogging:           debugLogging,
+		logger:                 logger,
+		augmentM3Tags:          agg.m3PrefixFilter,
 	}
-
-	return &downsampler{
-		opts:                opts.opts,
-		agg:                 opts.agg,
-		metricsAppenderOpts: metricsAppenderOpts,
-	}, nil
 }
 
 func (d *downsampler) NewMetricsAppender() (MetricsAppender, error) {
 	metricsAppender := d.agg.pools.metricsAppenderPool.Get()
+
+	d.RLock()
 	metricsAppender.reset(d.metricsAppenderOpts)
+	d.RUnlock()
+
 	return metricsAppender, nil
 }
+
+func (d *downsampler) OnUpdate(namespaces m3.ClusterNamespaces) {
+	logger := d.opts.InstrumentOptions.Logger()
+
+	if len(namespaces) == 0 {
+		logger.Debug("received empty list of namespaces. not updating staged metadata")
+		return
+	}
+
+	autoMappingRules, err := NewAutoMappingRules(namespaces)
+	if err != nil {
+		logger.Error("could not generate automapping rules for aggregated namespaces."+
+			" aggregations will continue with current configuration.", zap.Error(err))
+		return
+	}
+	var defaultStagedMetadatasProtos []metricpb.StagedMetadatas
```

Comment:
I think we can initialize w/ `len(autoMappingRules)`?

Reply:
+1